libs/java/instance_provider/src/main/java/com/yahoo/athenz/instance/provider/impl/InstanceZTSProvider.java

```diff
@@ -127,24 +157,113 @@ public InstanceConfirmation confirmInstance(InstanceConfirmation confirmation) {
+        if (sanIps == null || sanIps.length == 0) {
             return true;
         }
 
-        return (sanIPs.equals(clientIP));
+        if (clientIp == null || clientIp.isEmpty()) {
+            return false;
+        }
+
+        // It's possible both ipv4, ipv6 addresses are mentioned in sanIP
+        for (String sanIp: sanIps) {
+            if (sanIp.equals(clientIp)) {
+                return true;
+            }
+        }
+
+        LOGGER.error("Unable to match sanIp: {} with clientIp:{}", sanIps, clientIp);
+        return false;
     }
 
+    /**
+     * returns true if an empty hostname attribute is passed
+     * returns true if a non-empty hostname attribute is passed and all IPs passed in sanIp match the IPs that hostname resolves to.
+     * returns false in all other cases
+     * @param hostname
+     * @param sanIps
+     * @return true or false
+     */
+    boolean validateHostname(final String hostname, final String[] sanIps) {
+
+        LOGGER.debug("Validating hostname: {}, sanIps: {}", hostname, sanIps);
+
+        if (hostname == null || hostname.isEmpty()) {
+            LOGGER.info("Request contains no hostname entry for validation");
+            // if more than one sanIp is passed, all sanIPs must map to hostname, and hostname is a must
+            if (sanIps != null && sanIps.length > 1) {
+                LOGGER.error("SanIps:{} > 1, and hostname is empty", sanIps);
+                return false;
+            }
+            return true;
+        }
+
+        // IP in clientIp can be NATed. Rely on sanIp, which comes from the client, and is already matched with clientIp
+        // sanIp should be non-empty
+        if (sanIps == null || sanIps.length == 0) {
+            LOGGER.error("Request contains no sanIp entry for hostname:{} validation", hostname);
+            return false;
+        }
+
+        // All entries in sanIP must be one of the IPs that hostname resolves
+        Set<String>  hostIps = hostnameResolver.getAllByName(hostname);
+        for (String sanIp: sanIps) {
+            if (!hostIps.contains(sanIp)) {
+                LOGGER.error("One of sanIp: {} is not present in HostIps: {}", hostIps, sanIps);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * verifies if sanUri contains athenz://hostname/, the value matches the hostname
+     * @param sanUri
+     * @param hostname
+     * @return
+     */
+    boolean validateSanUri(final String sanUri, final String hostname) {
+
+        LOGGER.debug("Validating sanUri: {}, hostname: {}", sanUri, hostname);
+
+        if (sanUri == null || sanUri.isEmpty()) {
+            LOGGER.info("Request contains no sanURI to verify");
+            return true;
+        }
+
+        for (String uri: sanUri.split(",")) {
+            int idx = uri.indexOf(URI_HOSTNAME_PREFIX);
+            if (idx != -1) {
+                if (!uri.substring(idx + 18).equals(hostname)) {
```

Comment:
We should use URI_HOSTNAME_PREFIX.length() instead of hardcoded 18

Reply:
and that gets calculated on each api call, and for each uri. we are hardcoding known lengths in other places.

Reply:
ZTSProvider itself expected to be less frequently used. So, changed it to length().